<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Imports" data-toc-modified-id="Imports-0">Imports</a></span></li><li><span><a href="#Assemble-DataFrame" data-toc-modified-id="Assemble-DataFrame-1">Assemble DataFrame</a></span><ul class="toc-item"><li><span><a href="#Load-Files" data-toc-modified-id="Load-Files-1.1">Load Files</a></span></li><li><span><a href="#Review-Data" data-toc-modified-id="Review-Data-1.2">Review Data</a></span><ul class="toc-item"><li><span><a href="#Vizualize-DataFrames" data-toc-modified-id="Vizualize-DataFrames-1.2.1">Vizualize DataFrames</a></span></li><li><span><a href="#Data-Info" data-toc-modified-id="Data-Info-1.2.2">Data Info</a></span></li><li><span><a href="#Null-Values" data-toc-modified-id="Null-Values-1.2.3">Null Values</a></span></li><li><span><a href="#Column-Names" data-toc-modified-id="Column-Names-1.2.4">Column Names</a></span></li><li><span><a href="#Findings" data-toc-modified-id="Findings-1.2.5">Findings</a></span></li></ul></li><li><span><a href="#Prep-for-Concatenation" data-toc-modified-id="Prep-for-Concatenation-1.3">Prep for Concatenation</a></span><ul class="toc-item"><li><span><a href="#Rename-Columns" data-toc-modified-id="Rename-Columns-1.3.1">Rename Columns</a></span></li></ul></li><li><span><a href="#Concatenate-DataFrames" data-toc-modified-id="Concatenate-DataFrames-1.4">Concatenate DataFrames</a></span></li></ul></li><li><span><a href="#Data-Cleaning" data-toc-modified-id="Data-Cleaning-2">Data Cleaning</a></span><ul class="toc-item"><li><span><a href="#Remove-Nulls" data-toc-modified-id="Remove-Nulls-2.1">Remove Nulls</a></span></li><li><span><a href="#Remove-Duplicates" data-toc-modified-id="Remove-Duplicates-2.2">Remove Duplicates</a></span></li><li><span><a href="#'Description'-Clean-up" data-toc-modified-id="'Description'-Clean-up-2.3">'Description' Clean-up</a></span></li><li><span><a href="#'StockCode'-Clean-up" data-toc-modified-id="'StockCode'-Clean-up-2.4">'StockCode' Clean-up</a></span></li><li><span><a href="#'UnitPrice'-Clean-up" data-toc-modified-id="'UnitPrice'-Clean-up-2.5">'UnitPrice' Clean-up</a></span></li><li><span><a href="#'CustomerID'-Clean-up" data-toc-modified-id="'CustomerID'-Clean-up-2.6">'CustomerID' Clean-up</a></span></li><li><span><a href="#ID-Sales-vs.-Returns" data-toc-modified-id="ID-Sales-vs.-Returns-2.7">ID Sales vs. Returns</a></span></li><li><span><a href="#'InvoiceNo'-Clean-up" data-toc-modified-id="'InvoiceNo'-Clean-up-2.8">'InvoiceNo' Clean-up</a></span></li><li><span><a href="#Eliminate-Returns-for-Undocumented-Sales" data-toc-modified-id="Eliminate-Returns-for-Undocumented-Sales-2.9">Eliminate Returns for Undocumented Sales</a></span></li><li><span><a href="#'InvoiceDate'-Clean-Up" data-toc-modified-id="'InvoiceDate'-Clean-Up-2.10">'InvoiceDate' Clean-Up</a></span></li></ul></li><li><span><a href="#Save-to-CSV" data-toc-modified-id="Save-to-CSV-3">Save to CSV</a></span></li></ul></div>

# Imports

In [275]:
import pandas as pd
import numpy as np
import re
import string
from IPython.display import clear_output

# Assemble DataFrame

**The sales data will be compiled from two spreadsheets containing online sales records from an online retailer of homegoods in the UK.**  

**Files were downloaded from University California Irvine's Machine Learning Repository:**
1. http://archive.ics.uci.edu/ml/datasets/Online+Retail  
2. http://archive.ics.uci.edu/ml/datasets/Online+Retail+II


**Files were concatenated and any overlapping records were dropped.**

## Load Files

In [20]:
# Load first spreadsheet
sales_data_a = pd.read_excel('online_retail_II.xlsx')

In [21]:
# Load second spreadsheet
sales_data_b = pd.read_excel('Online Retail.xlsx')

## Review Data

### Vizualize DataFrames

In [22]:
# Visualize dataframe a sorted by InvoiceDate
sales_data_a.sort_values('InvoiceDate', ascending = True).head()

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom

In [23]:
# Visualize last rows of dataframe a sorted by InvoiceDate
sales_data_a.sort_values('InvoiceDate', ascending = True).tail()

Invoice StockCode             Description  Quantity  \
525423  538171     20771  CHRYSANTHEMUM  JOURNAL         1   
525424  538171     21882             SKULLS TAPE         5   
525425  538171     21879        HEARTS GIFT TAPE         5   
525411  538171     20777  CHRYSANTHEMUM NOTEBOOK         2   
525460  538171     21931  JUMBO STORAGE BAG SUKI         2   

               InvoiceDate  Price  Customer ID         Country  
525423 2010-12-09 20:01:00   2.55      17530.0  United Kingdom  
525424 2010-12-09 20:01:00   0.65      17530.0  United Kingdom  
525425 2010-12-09 20:01:00   0.65      17530.0  United Kingdom  
525411 2010-12-09 20:01:00   1.65      17530.0  United Kingdom  
525460 2010-12-09 20:01:00   1.95      17530.0  United Kingdom

In [24]:
# Visualize dataframe b sorted by InvoiceDate
sales_data_b.sort_values('InvoiceDate', ascending = True).head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [25]:
# Visualize last rows of dataframe b sorted by InvoiceDate
sales_data_b.sort_values('InvoiceDate', ascending = True).tail()

InvoiceNo StockCode                      Description  Quantity  \
541896    581587     22555        PLASTERS IN TIN STRONGMAN        12   
541895    581587     22556   PLASTERS IN TIN CIRCUS PARADE         12   
541907    581587     23255  CHILDRENS CUTLERY CIRCUS PARADE         4   
541900    581587     22730       ALARM CLOCK BAKELIKE IVORY         4   
541908    581587     22138    BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID Country  
541896 2011-12-09 12:50:00       1.65     12680.0  France  
541895 2011-12-09 12:50:00       1.65     12680.0  France  
541907 2011-12-09 12:50:00       4.15     12680.0  France  
541900 2011-12-09 12:50:00       3.75     12680.0  France  
541908 2011-12-09 12:50:00       4.95     12680.0  France

### Data Info

In [26]:
# Info for dataframe a
sales_data_a.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
Invoice        525461 non-null object
StockCode      525461 non-null object
Description    522533 non-null object
Quantity       525461 non-null int64
InvoiceDate    525461 non-null datetime64[ns]
Price          525461 non-null float64
Customer ID    417534 non-null float64
Country        525461 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 32.1+ MB


In [27]:
# # Info for dataframe b
sales_data_b.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
InvoiceNo      541909 non-null object
StockCode      541909 non-null object
Description    540455 non-null object
Quantity       541909 non-null int64
InvoiceDate    541909 non-null datetime64[ns]
UnitPrice      541909 non-null float64
CustomerID     406829 non-null float64
Country        541909 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


### Null Values

In [28]:
# ID columns in dataframe a with null values
sales_data_a.isnull().any().sort_values(ascending = False)

Customer ID     True
Description     True
Country        False
Price          False
InvoiceDate    False
Quantity       False
StockCode      False
Invoice        False
dtype: bool

In [29]:
# Show % of nulls by column in dataframe a
(sales_data_a.isnull().sum() / sales_data_a.shape[0]).sort_values(ascending = False)

Customer ID    0.205395
Description    0.005572
Country        0.000000
Price          0.000000
InvoiceDate    0.000000
Quantity       0.000000
StockCode      0.000000
Invoice        0.000000
dtype: float64

In [30]:
# ID columns in dataframe b with null values
sales_data_b.isnull().any().sort_values(ascending = False)

CustomerID      True
Description     True
Country        False
UnitPrice      False
InvoiceDate    False
Quantity       False
StockCode      False
InvoiceNo      False
dtype: bool

In [31]:
# Show % of nulls by column in dataframe b
(sales_data_b.isnull().sum() / sales_data_b.shape[0]).sort_values(ascending = False)

CustomerID     0.249267
Description    0.002683
Country        0.000000
UnitPrice      0.000000
InvoiceDate    0.000000
Quantity       0.000000
StockCode      0.000000
InvoiceNo      0.000000
dtype: float64

### Column Names

In [32]:
# Column names in dataframe a
sales_data_a.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='object')

In [33]:
# Column names in dataframe b
sales_data_b.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

### Findings

**1. Both dataframes have overlapping rows from:**    
- InvoiceNo 536365 to 538171    
- InvoiceDates from 2010-12-01 08:28:00 to 2010-12-09 20:01:00
  
**2. CustomerID and Description columns have null values**  
**3. Column names do not match between dataframes**

## Prep for Concatenation

### Rename Columns

**Rename columns in dataframe a to match those in b to facilitate concatenation.**

In [34]:
# Rename columns in dataframe a
sales_data_a = sales_data_a.rename(columns = {'Invoice':'InvoiceNo', 
                                              'Price':'UnitPrice', 
                                              'Customer ID':'CustomerID'})

In [35]:
# Confirm column names match between dataframes
assert sales_data_a.columns.all() == sales_data_b.columns.all()

## Concatenate DataFrames

In [387]:
# Concatenate sales_data_a with sales_data_b
all_data = pd.concat([sales_data_a, sales_data_b])

In [388]:
# View all_data info
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1067370 entries, 0 to 541908
Data columns (total 8 columns):
InvoiceNo      1067370 non-null object
StockCode      1067370 non-null object
Description    1062988 non-null object
Quantity       1067370 non-null int64
InvoiceDate    1067370 non-null datetime64[ns]
UnitPrice      1067370 non-null float64
CustomerID     824363 non-null float64
Country        1067370 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 73.3+ MB


# Data Cleaning

## Remove Nulls

In [389]:
# Drop all rows with null values
all_data = all_data.dropna(how = 'any', axis = 0)

In [390]:
# Look dataframe info to confirm no nulls remain
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 824363 entries, 0 to 541908
Data columns (total 8 columns):
InvoiceNo      824363 non-null object
StockCode      824363 non-null object
Description    824363 non-null object
Quantity       824363 non-null int64
InvoiceDate    824363 non-null datetime64[ns]
UnitPrice      824363 non-null float64
CustomerID     824363 non-null float64
Country        824363 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 56.6+ MB


## Remove Duplicates

In [391]:
# Create a boolean column, 'duplicates', that will be True if row is a duplicate and False if not
# Duplicates will be determined using subset of columns entered
# Description not included in subset columns because there are errors present that results in duplicates being flagged as unique
# keep = 'first': marks instances after first as 'True'
all_data['duplicates'] = all_data.duplicated(subset = ['StockCode', 
                                                       'InvoiceNo', 
                                                       'InvoiceDate', 
                                                       'CustomerID', 
                                                       'Quantity', 
                                                       'UnitPrice'], keep = 'first')

In [392]:
# Keep rows where duplicates column value was False (aka non-duplicate rows) and drop duplicates column
all_data = (all_data[all_data.duplicates == False].sort_values('InvoiceDate', ascending = True)
                                                 .drop(['duplicates'], axis = 1).reset_index(drop=True))

In [393]:
# all_data info after duplicates were dropped
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 797882 entries, 0 to 797881
Data columns (total 8 columns):
InvoiceNo      797882 non-null object
StockCode      797882 non-null object
Description    797882 non-null object
Quantity       797882 non-null int64
InvoiceDate    797882 non-null datetime64[ns]
UnitPrice      797882 non-null float64
CustomerID     797882 non-null float64
Country        797882 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 48.7+ MB


## 'Description' Clean-up

**1. Drop rows were Description is 'Manual' and no product name.**  

In [394]:
# Keep only rows where Description is not 'Manual'
all_data = all_data[all_data['Description'] != 'Manual'].reset_index(drop = True)

In [395]:
# all_data info after Descriptions of 'Manual' were dropped
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 796797 entries, 0 to 796796
Data columns (total 8 columns):
InvoiceNo      796797 non-null object
StockCode      796797 non-null object
Description    796797 non-null object
Quantity       796797 non-null int64
InvoiceDate    796797 non-null datetime64[ns]
UnitPrice      796797 non-null float64
CustomerID     796797 non-null float64
Country        796797 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 48.6+ MB


**2. Remove spaces before and after descriptions and remove non-alphanumeric characters.**

In [396]:
# Strip spaces before and after descriptions
all_data.Description = all_data.Description.apply(lambda x: x.strip())

In [397]:
# Replace any non-alphanumeric character with a space
all_data.Description = all_data.Description.apply(lambda x: re.sub('[^0-9a-zA-Z\' ]+', ' ', x))

In [398]:
# Replace multiple spaces with one space
all_data.Description = all_data.Description.apply(lambda x: re.sub(' +', ' ', x))

In [399]:
# Visualize all descriptions
for x in all_data.Description.sort_values(ascending = True).unique():
    print(x)

10 COLOUR SPACEBOY PEN
11 PC CERAMIC TEA SET POLKADOT
12 ASS ZINC CHRISTMAS DECORATIONS
12 COLOURED PARTY BALLOONS
12 DAISY PEGS IN WOOD BOX
12 EGG HOUSE PAINTED WOOD
12 HANGING EGGS HAND PAINTED
12 IVORY ROSE PEG PLACE SETTINGS
12 MESSAGE CARDS WITH ENVELOPES
12 MINI TOADSTOOL PEGS
12 PENCIL SMALL TUBE WOODLAND
12 PENCILS SMALL TUBE POSY
12 PENCILS SMALL TUBE RED RETROSPOT
12 PENCILS SMALL TUBE RED SPOTTY
12 PENCILS SMALL TUBE SKULL
12 PENCILS TALL TUBE POSY
12 PENCILS TALL TUBE RED RETROSPOT
12 PENCILS TALL TUBE RED SPOTTY
12 PENCILS TALL TUBE SKULLS
12 PENCILS TALL TUBE WOODLAND
12 PINK HEN CHICKS IN BASKET
12 PINK ROSE PEG PLACE SETTINGS
12 RED ROSE PEG PLACE SETTINGS
15 PINK FLUFFY CHICKS IN BOX
15CM CHRISTMAS GLASS BALL 20 LIGHTS
16 PC CUTLERY SET PANTRY DESIGN
16 PIECE CUTLERY SET PANTRY DESIGN
18PC WOODEN CUTLERY SET DISPOSABLE
2 DAISIES HAIR COMB
2 PICTURE BOOK EGGS EASTER BUNNY
2 PICTURE BOOK EGGS EASTER CHICKS
2 PICTURE BOOK EGGS EASTER DUCKS
20 DOLLY PEGS RETROSPOT
200 BEND

FLOWER PURPLE CLOCK W SUCKER
FLOWER PURPLE CLOCK WITH SUCKER
FLOWER SHOP DESIGN MUG
FLOWER VINE RAFFIA FOOD COVER
FLOWER WINDMILL
FLOWERS AND BIRDS WALL PLAQUE
FLOWERS CHANDELIER T LIGHT HOLDER
FLOWERS HANDBAG blue and orange
FLOWERS STICKERS
FLOWERS TILE COASTER
FLOWERS TILE HOOK
FLOWERS WALL THERMOMETER
FLUTED ANTIQUE CANDLE HOLDER
FLYING PIG WATERING CAN
FLYING SAUCER TABLE LIGHT IN WHITE
FOLDING BUTTERFLY MIRROR HOT PINK
FOLDING BUTTERFLY MIRROR IVORY
FOLDING BUTTERFLY MIRROR RED
FOLDING CAMPING SCISSOR W KNIF S
FOLDING MIRROR HOT PINK
FOLDING MIRROR IVORY
FOLDING MIRROR RED
FOLDING SHIRT TIDY
FOLDING SHOE TIDY
FOLDING UMBRELLA BLACK BLUE SPOT
FOLDING UMBRELLA BLACKBLUE POLKADOT
FOLDING UMBRELLA CHOC WHITE SPOT
FOLDING UMBRELLA CHOCOLATE POLKADOT
FOLDING UMBRELLA CREAM MULTI SPOT
FOLDING UMBRELLA CREAM POLKADOT
FOLDING UMBRELLA PINK WHITE SPOT
FOLDING UMBRELLA PINKWHITE POLKADOT
FOLDING UMBRELLA RED WHITE POLKADOT
FOLDING UMBRELLA RED WHITE SPOT
FOLDING UMBRELLA WHITE RED POLKADOT


MINI JIGSAW BAKE A CAKE
MINI JIGSAW BUNNIES
MINI JIGSAW BUNNIES AT HOME
MINI JIGSAW CIRCUS PARADE
MINI JIGSAW DINOSAUR
MINI JIGSAW DOLLY GIRL
MINI JIGSAW FIDO
MINI JIGSAW GO TO THE FAIR
MINI JIGSAW LEAP FROG
MINI JIGSAW PURDEY
MINI JIGSAW SPACEBOY
MINI LADLE LOVE HEART PINK
MINI LADLE LOVE HEART RED
MINI LIGHTS WOODLAND MUSHROOMS
MINI PAINT SET VINTAGE
MINI PAINT SET VINTAGE BOY GIRL
MINI PAINTED GARDEN DECORATION
MINI PLAYING CARDS BUFFALO BILL
MINI PLAYING CARDS DOLLY GIRL
MINI PLAYING CARDS FUN FAIR
MINI PLAYING CARDS GYMKHANA
MINI PLAYING CARDS SPACEBOY
MINI WOODEN HAPPY BIRTHDAY GARLAND
MINI ZINC GARDEN DECORATIONS
MINIATURE ANTIQUE ROSE HOOK IVORY
MINITURE ANTIQUE ROSE HOOK IVORY
MINT DINER CLOCK
MINT DINER WALL CLOCK
MINT GREEN ROSE TOWEL
MINT KITCHEN SCALES
MIRROR ARCHED GEORGIAN
MIRROR CORNICE
MIRROR LOVE BIRD T LIGHT HOLDER
MIRROR MOSAIC CANDLE PLATE
MIRROR MOSAIC GOBLET CANDLE HOLDER
MIRROR MOSAIC HURRICANE LAMP
MIRROR MOSAIC T LIGHT HOLDER
MIRROR MOSAIC T LIGHT HOLDER ROUND

SET OF 4 KNICK KNACK TINS LEAVES
SET OF 4 KNICK KNACK TINS LONDON
SET OF 4 KNICK KNACK TINS POPPIES
SET OF 4 NAPKIN CHARM CUTLERY
SET OF 4 NAPKIN CHARMS 3 KEYS
SET OF 4 NAPKIN CHARMS CROWNS
SET OF 4 NAPKIN CHARMS CUTLERY
SET OF 4 NAPKIN CHARMS HEARTS
SET OF 4 NAPKIN CHARMS INSTRUMENT
SET OF 4 NAPKIN CHARMS INSTUMENT
SET OF 4 NAPKIN CHARMS LEAVES
SET OF 4 NAPKIN CHARMS STARS
SET OF 4 NEW ENGLAND COASTERS
SET OF 4 NEW ENGLAND PLACEMATS
SET OF 4 PANTRY JELLY MOULDS
SET OF 4 PASTEL STACKING MUGS
SET OF 4 PISTACHIO LOVEBIRD COASTER
SET OF 4 POLKADOT COASTERS
SET OF 4 POLKADOT PLACEMATS
SET OF 4 POLKADOTS COASTERS
SET OF 4 ROSE BOTANICAL CANDLES
SET OF 4 SANTA PLACE SETTINGS
SET OF 5 LUCKY CAT MAGNETS
SET OF 5 MINI GROCERY MAGNETS
SET OF 5 PANCAKE DAY MAGNETS
SET OF 6 3D KIT CARDS FOR KIDS
SET OF 6 CAKE CHOPSTICKS
SET OF 6 DOTS CHOPSTICKS
SET OF 6 EASTER RAINBOW CHICKS
SET OF 6 FUNKY BEAKERS
SET OF 6 GIRLS CELEBRATION CANDLES
SET OF 6 HALLOWEEN GHOST T LIGHTS
SET OF 6 HEART CHOPSTICKS
SET OF

**3. Replace Descriptions for StockCodes with multiple descriptions with a uniform description.**

In [400]:
def count_desc(df):
    '''Function finds and counts StockCodes with more than one Description.  Function returns the number of
       StockCodes with duplicate Descriptions and a list of the codes.  
       Inputs:
         df = dataframe to reference
         code_count = counter to record number of StockCodes with > 1 Description
         mult_desc_codes = empty list initiated to record StockCodes with multiple descriptions
       Outputs:
         code_count and mult_desc_codes
     '''
    code_series = df.groupby(['StockCode']).Description.nunique().sort_values(ascending = False)
    code_count = 0
    mult_desc_codes = []
    
    # For each StockCode, increment code_count and record StockCode to list if number of unique Description > 1 
    for row in range(0,len(code_series)):
        if code_series.iloc[row] > 1:
            code_count += 1
            mult_desc_codes.append(code_series.index[row])
    return code_count, mult_desc_codes

In [401]:
# Run count_desc() function and store outputs
code_count, mult_desc_codes = count_desc(all_data)
code_count, mult_desc_codes

(534,
 [23236,
  22384,
  20685,
  23196,
  23244,
  22345,
  '84997A',
  23209,
  23535,
  22191,
  23413,
  22785,
  22356,
  23131,
  22937,
  23366,
  22344,
  21524,
  21523,
  23370,
  '84509C',
  23231,
  '85099B',
  21955,
  22776,
  22346,
  '84997C',
  22333,
  '84997D',
  22139,
  22536,
  22952,
  22844,
  22950,
  23396,
  22416,
  '17107D',
  23126,
  'ADJUST',
  22845,
  22852,
  23203,
  '84997B',
  22197,
  22740,
  22853,
  21844,
  21398,
  22492,
  21399,
  22892,
  22897,
  72586,
  21080,
  '37482P',
  21397,
  22343,
  22896,
  '85123A',
  21843,
  22134,
  22347,
  22348,
  22843,
  21157,
  21156,
  21155,
  21154,
  84969,
  21811,
  22367,
  22366,
  22365,
  23691,
  21124,
  21123,
  21122,
  21121,
  21818,
  21391,
  22359,
  21392,
  21393,
  21394,
  22898,
  22353,
  22352,
  '82001S',
  22350,
  21395,
  22354,
  22328,
  22899,
  72817,
  22939,
  20983,
  22511,
  '72800D',
  23497,
  21899,
  '72800C',
  22903,
  '72800B',
  23502,
  72814,
  22659

In [402]:
# Replace alternate descriptions with the first description used
for code in mult_desc_codes:
    descriptions = list(all_data[all_data.StockCode == code].Description.unique())
    master_desc = descriptions[0]
    
    if len(descriptions) > 1:
        # clears printed values
        clear_output(wait=True)
        
        # print current status
        print(f'\rCurrently replacing: {descriptions}\n\twith {master_desc}', end="\r", flush=True)
        
        # replace description
        for versions in descriptions[1:]:
            all_data.Description = all_data.Description.replace(versions, master_desc)
    else:
        continue
        
# check that no StockCodes have multiple descriptions
if all_data.groupby(['StockCode']).Description.nunique().sort_values(ascending = False).max() == 1:
    clear_output(wait=True)
    print(f'Finished!  All duplicate descriptions have been removed')
else:
    print(f'NOT COMPLETE!')

Finished!  All duplicate descriptions have been removed


In [403]:
# Double check that each StockCode only has 1 description
assert all_data.groupby(['StockCode']).Description.nunique().max() == 1

## 'StockCode' Clean-up

**1. Remove rows where StockCode was for a test item or an adjustment.**

In [404]:
# Initiate list to store rows indices with descriptions starting with 'TEST' or 'ADJUST' 
drop_index = []

# Define regex pattern for finding StockCodes that start with 'TEST' or 'ADJUST'
desc_pattern = re.compile('^(TEST|ADJUST)') 

# For each row in dataframe, check if regex pattern is found and store index to drop_index if it is
for i in range(0,all_data.shape[0]):
    try:
        result = desc_pattern.match(all_data.StockCode.iloc[i])
        if result:
            drop_index.append(i)
    except: 
        if type(all_data.StockCode.iloc[i]) == int:
            continue

In [405]:
# Visualize drop_index list
print([x for x in drop_index])

[21433, 21682, 21691, 28475, 28487, 31256, 31258, 37891, 44868, 48912, 48913, 48914, 48915, 48916, 48917, 48918, 48919, 48920, 48921, 48922, 48923, 48935, 48936, 48967, 48968, 48969, 48970, 48971, 48972, 48973, 48974, 48975, 48976, 48977, 48978, 48979, 48980, 48981, 48982, 48984, 48985, 48986, 48987, 48988, 48989, 48990, 48991, 48992, 48993, 48994, 48995, 48996, 48997, 48998, 48999, 49000, 49001, 49002, 49003, 49004, 49005, 49006, 49007, 49008, 49009, 49010, 49011, 62866, 62954, 116167, 159197, 159198, 159199, 191565, 191566, 191569, 291123, 291124, 291125, 291126]


In [406]:
# Drop rows with StockCodes for test item or adjustment
all_data = all_data.drop(index = drop_index)

In [407]:
# all_data info after drop
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 796717 entries, 0 to 796796
Data columns (total 8 columns):
InvoiceNo      796717 non-null object
StockCode      796717 non-null object
Description    796717 non-null object
Quantity       796717 non-null int64
InvoiceDate    796717 non-null datetime64[ns]
UnitPrice      796717 non-null float64
CustomerID     796717 non-null float64
Country        796717 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 54.7+ MB


**2. Remove StockCodes corresponding to discounts given to customers.**

In [408]:
# StockCode D represents discounts given to customers
all_data[all_data.StockCode == 'D'].head()

InvoiceNo StockCode Description  Quantity         InvoiceDate  \
716     C489535         D    Discount        -1 2009-12-01 12:11:00   
717     C489535         D    Discount        -1 2009-12-01 12:11:00   
18228   C491428         D    Discount        -1 2009-12-10 20:23:00   
23296   C491962         D    Discount        -1 2009-12-14 16:38:00   
28286   C492693         D    Discount        -1 2009-12-17 18:47:00   

       UnitPrice  CustomerID         Country  
716         9.00     15299.0  United Kingdom  
717        19.00     15299.0  United Kingdom  
18228       9.10     15494.0  United Kingdom  
23296       0.59     13206.0  United Kingdom  
28286       6.85     13408.0  United Kingdom

In [409]:
# Remove rows with StockCode == D
all_data = all_data.loc[all_data.StockCode != 'D'].reset_index(drop = True)

In [410]:
# all_data info after discounts were dropped
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 796547 entries, 0 to 796546
Data columns (total 8 columns):
InvoiceNo      796547 non-null object
StockCode      796547 non-null object
Description    796547 non-null object
Quantity       796547 non-null int64
InvoiceDate    796547 non-null datetime64[ns]
UnitPrice      796547 non-null float64
CustomerID     796547 non-null float64
Country        796547 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 48.6+ MB


**Find StockCodes that start with non-numeric characters since they may indicate non-product items**

In [411]:
# Check each unique StockCode for non-numeric values and print them
for x in all_data.StockCode.unique():
    if type(x) == str and re.match('^\D',x): 
        print(x)
    else:
        continue

POST
C2
BANK CHARGES
PADS
SP1002
DOT
CRUK


**Items that are not part of inventory**  
- POST:         Postage for shipping order  
- BANK CHARGES: Processing fees  
- PADS:         Possibly packing supplies  
- DOT:          Dotcom postage   
- CRUK:         Commision charges

**3. Remove StockCodes that are not inventory**

In [412]:
# Keep only StockCodes that are inventory
all_data = all_data.loc[(all_data.StockCode!='POST') &
                    (all_data.StockCode!='BANK CHARGES') &
                    (all_data.StockCode!='PADS') &
                    (all_data.StockCode!='DOT') &
                    (all_data.StockCode!='CRUK')].reset_index(drop = True)

In [413]:
# all_data info after dropping non-inventory StockCodes
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 794477 entries, 0 to 794476
Data columns (total 8 columns):
InvoiceNo      794477 non-null object
StockCode      794477 non-null object
Description    794477 non-null object
Quantity       794477 non-null int64
InvoiceDate    794477 non-null datetime64[ns]
UnitPrice      794477 non-null float64
CustomerID     794477 non-null float64
Country        794477 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 48.5+ MB


**There are also items with multiple StockCodes**

In [414]:
# Check which Descriptions are linked to multiple StockCodes
multi_code = all_data.groupby(['Description']).StockCode.nunique().sort_values(ascending = False)
multi_code

Description
METAL SIGN CUPCAKE SINGLE HOOK    4
COLUMBIAN CANDLE ROUND            4
MODERN CHRISTMAS TREE CANDLE      3
COLOURING PENCILS BROWN TUBE      3
WHITE BAMBOO RIBS LAMPSHADE       3
                                 ..
POLKA DOT RAFFIA FOOD COVER       1
POLKA DOT PISTACHIO PAIL          1
POLKA DOT PEN                     1
POCKET MIRROR WOODLAND            1
10 COLOUR SPACEBOY PEN            1
Name: StockCode, Length: 4587, dtype: int64

In [415]:
# Create list of Descriptions with more than 1 StockCode
multi_code_items = list(multi_code[multi_code>1].index.get_level_values(0))
multi_code_items

['METAL SIGN CUPCAKE SINGLE HOOK',
 'COLUMBIAN CANDLE ROUND',
 'MODERN CHRISTMAS TREE CANDLE',
 'COLOURING PENCILS BROWN TUBE',
 'WHITE BAMBOO RIBS LAMPSHADE',
 'FRENCH LATTICE CUSHION COVER',
 'BROCANTE SHELF WITH HOOKS',
 'HEART T LIGHT HOLDER',
 'FAIRY CAKE PLACEMATS',
 'PINK JEWELLED PHOTO FRAME',
 'ROSE DU SUD CUSHION COVER',
 'EAU DE NILE JEWELLED PHOTOFRAME',
 'CHARLIE LOLA RED HOT WATER BOTTLE',
 "PINK B'FLY C COVER W BOBBLES",
 'WATERING CAN GREEN DINOSAUR',
 'CHERRY BLOSSOM DECORATIVE FLASK',
 'ROUND ARTICULATED BLUE CLOCK W SUCK',
 'BLUE FLOCK GLASS CANDLEHOLDER',
 'PASTEL BLUE PHOTO ALBUM',
 'PASTEL PINK PHOTO ALBUM',
 'SQUARE CHERRY BLOSSOM CABINET',
 'FROSTED WHITE BASE',
 'ACRYLIC JEWEL SNOWFLAKE BLUE',
 'ACRYLIC JEWEL SNOWFLAKE PINK',
 'COLUMBIAN CANDLE RECTANGLE',
 'COLUMBIAN CUBE CANDLE',
 'RETRO PLASTIC POLKA TRAY',
 'SILVER VANILLA FLOWER CANDLE POT',
 'RETRO PLASTIC DAISY TRAY',
 "RETRO PLASTIC 70'S TRAY",
 'FRENCH PAISLEY CUSHION COVER',
 'PAPER LANTERN 9 POINT SN

**4. Replace StockCodes for items with multiple StockCodes with the mode StockCode**

In [416]:
# For each Description with > 1 StockCode, replace the StockCode mode StockCode for that Description
for item in multi_code_items:
    all_data.loc[all_data.Description == item,'StockCode'] = all_data[all_data.Description == item].StockCode.mode()[0]

In [417]:
# Check that there are no Descriptions with multiple StockCodes
all_data.groupby(['Description']).StockCode.nunique().sort_values(ascending = False).max()

1

In [418]:
# Check that there are no StockCodes with multilple Descriptions
all_data.groupby(['StockCode']).Description.nunique().sort_values(ascending = False).max()

1

In [419]:
# Check that the number of unique StockCodes matches number of unique Descriptions
assert all_data.groupby(['Description']).StockCode.nunique().sum() == all_data.groupby(['StockCode']).Description.nunique().sum()

## 'UnitPrice' Clean-up

In [420]:
# Count number of UnitPrices for each product Description
p_df = pd.DataFrame(all_data.groupby(['Description']).UnitPrice.nunique().sort_values(ascending = False))
p_df.head()

UnitPrice
Description                         
COOK WITH WINE METAL SIGN         13
DOOR MAT UNION FLAG               12
RED SPOTTY COIR DOORMAT           12
DOOR MAT WELCOME PUPPIES          12
DOOR MAT FAIRY CAKE               12

**1. Drop rows where UnitPrice was zero**

In [421]:
# drop rows where items were sold at 0 cost
all_data = all_data.loc[(all_data.UnitPrice > 0.00)]

In [422]:
# Create df with Unit price stats
p_df['min'] = all_data.groupby(['Description']).UnitPrice.min()
p_df['max'] = all_data.groupby(['Description']).UnitPrice.max()
p_df['range'] = p_df['max']-p_df['min']
p_df_sorted = p_df.sort_values('range', ascending = False)
p_df_sorted

UnitPrice    min     max  range
Description                                                      
PICNIC BASKET WICKER SMALL                6   2.00  649.50  647.5
VINTAGE RED KITCHEN CABINET               4  50.00  295.00  245.0
VINTAGE BLUE KITCHEN CABINET              6  50.00  295.00  245.0
LOVE SEAT ANTIQUE WHITE METAL             4  42.50  195.00  152.5
CARRIAGE                                  6  18.00  150.00  132.0
...                                     ...    ...     ...    ...
NEW ENGLAND EGG WARMER                    1   0.55    0.55    0.0
NEW ENGLAND MILK JUG W GIFT BOX           1   5.95    5.95    0.0
ASSORTED TUTTI FRUTTI NOTEBOOK            1   2.10    2.10    0.0
NEW ENGLAND MUG W GIFT BOX                1   2.55    2.55    0.0
FOLDING CAMPING SCISSOR W KNIF S          1   0.12    0.12    0.0

[4587 rows x 4 columns]

In [423]:
# # Check mode compared to median and max/min to see if changing all prices to mode will eliminate outliers
# for desc in p_df_sorted.index.get_level_values(0):
#     print(f'{desc}')
#     print(f'\tmedian: {all_data[all_data.Description == desc].UnitPrice.median()}')
#     print(f'\tmode: {all_data[all_data.Description == desc].UnitPrice.mode()[0]}')
#     print(f'\tmin: {all_data[all_data.Description == desc].UnitPrice.min()}')
#     print(f'\tmax: {all_data[all_data.Description == desc].UnitPrice.max()}\n')
#     print(f'\tCounts: \n{all_data[all_data.Description == desc].UnitPrice.value_counts()}\n')

For items with multiple unit prices, the mode seems to be a good representation of price.   
**2. Replace price for items with multiple UnitPrices with their mode.**

In [424]:
# Create list of items with multiple prices
multi_price_items = list(p_df_sorted[p_df_sorted.UnitPrice>1].index.get_level_values(0))

In [425]:
# Replace the price of each item with the mode price
for item in multi_price_items:
    all_data.loc[all_data.Description == item,'UnitPrice'] = all_data[all_data.Description == item].UnitPrice.mode()[0]

In [426]:
# Check to ensure no Descriptions have more than 1 UnitPrice
all_data.groupby(['Description']).UnitPrice.nunique().max()

1

In [427]:
# Reset index and show all_data info after UnitPrices were cleaned up
all_data.reset_index(drop=True).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 794417 entries, 0 to 794416
Data columns (total 8 columns):
InvoiceNo      794417 non-null object
StockCode      794417 non-null object
Description    794417 non-null object
Quantity       794417 non-null int64
InvoiceDate    794417 non-null datetime64[ns]
UnitPrice      794417 non-null float64
CustomerID     794417 non-null float64
Country        794417 non-null object
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 48.5+ MB


**Check UnitPrice range across all products**

In [428]:
print(f'The most expensive product costs ${all_data.UnitPrice.max()} per unit')
print(f'The least expensive product costs ${all_data.UnitPrice.min()} per unit')

The most expensive product costs $295.0 per unit
The least expensive product costs $0.04 per unit


## 'CustomerID' Clean-up

**Change datatype to integer**

In [429]:
all_data.CustomerID = all_data.CustomerID.astype(int)

In [430]:
# all_data info after data type change
all_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 794417 entries, 0 to 794476
Data columns (total 8 columns):
InvoiceNo      794417 non-null object
StockCode      794417 non-null object
Description    794417 non-null object
Quantity       794417 non-null int64
InvoiceDate    794417 non-null datetime64[ns]
UnitPrice      794417 non-null float64
CustomerID     794417 non-null int64
Country        794417 non-null object
dtypes: datetime64[ns](1), float64(1), int64(2), object(4)
memory usage: 54.5+ MB


## ID Sales vs. Returns

**1. Specify type as 'sale' for transactions with InvoiceNo not beginning with 'C', aka transactions with Quantities > 0**

In [431]:
all_data.loc[all_data['InvoiceNo'].str[0]!='C', 'Type'] = 'sale'

**2. Specify type as 'return' for transactions with InvoiceNo beginning with 'C'**

In [432]:
all_data.loc[all_data['InvoiceNo'].str[0]=='C', 'Type'] = 'return'

## 'InvoiceNo' Clean-up

**1. Remove 'C' from InvoiceNo for 'returns'**

In [433]:
all_data.loc[all_data['InvoiceNo'].str[0]=='C', 'InvoiceNo'] = all_data['InvoiceNo'].str[1:]

**2. Convert InvoiceNo data to integers**

In [434]:
all_data['InvoiceNo'] = all_data['InvoiceNo'].astype(int)

## Eliminate Returns for Undocumented Sales

**Returns that are not preceeded by sales that account for the return quantity will be excluded to prevent negative total sales quantities** 

**Generate list of tuples representing all returns**

In [361]:
# Set up mask
return_mask = all_data[all_data['Type']== 'return'].groupby(['CustomerID', 'StockCode', 'InvoiceDate']).sum().reset_index()
check_list = []
return_mask

# for each stockcode customer combo with a return
# find all sales for that combo occuring before date of last retrun
# if sum of all sales qty < sum of return qty then drop return
for row in range(0,return_mask.shape[0]):
    check_list.append((return_mask.loc[row,'CustomerID'],
                       return_mask.loc[row,'StockCode'],
                       return_mask.loc[row,'InvoiceDate']))

**Assemble list of indices to drop**

In [362]:
index_to_drop = []
count = 0

# For each item in the checklist, add its index to index_to_drop list if 
for item in check_list:
    
    # assign variables to tuple items
    custID = item[0]
    stockCo = item[1]
    rowDate = item[2]

    # create mask for that customer and StockCode where transaction is a return
    cust_mask = (all_data[(all_data['CustomerID'] == custID) & 
                          (all_data['StockCode'] == stockCo) & 
                          (all_data['Type'] == 'return')])
    
    # count number of unique InvoiceDates
    n_dates = cust_mask.InvoiceDate.nunique()
    
    # list unique InvoiceDates
    unique_dates = cust_mask.InvoiceDate.unique()
    
    # sum all returns for specific StockCode and CustomerID
    return_qty = cust_mask.Quantity.sum()
    
    # ID most recent return of specific product
    return_date = max(unique_dates)
    
    # create a mask to show all purchases by the customer of this specific product before the date of the last return
    sale_mask = (all_data[(all_data['CustomerID'] == custID) & 
                               (all_data['StockCode'] == stockCo) & 
                               (all_data['InvoiceDate'] <= return_date) &
                               (all_data['Type'] == 'sale')])
    
    # sum all purchases of the specific item up to the most recent return date
    buy_qty_to_date = sale_mask.Quantity.sum()

    # subtract qty returned (negative number) from qty purchased to get net_sale qty for the item/customer pair
    net_sale = buy_qty_to_date + return_qty
    
    # if the return date = the most recent return of this item and the qty returned is more than what has been purchased in record
    # add all invoice dates for that item/customer pair to index_to_drop list
    # add all sales dates for that item/customer pair to index_to_drop list
    if rowDate == return_date and net_sale < 0:
        return_idx = cust_mask[cust_mask['InvoiceDate'] <= return_date].index
        sale_idx = sale_mask[sale_mask['InvoiceDate'] <= return_date].index
        
        index_to_drop.extend(return_idx)
        index_to_drop.extend(sale_idx)

In [382]:
# Number of indices to drop
len(index_to_drop)

1974

**1. Drop CustomerID/StockCode pairs that have returns and sales that add up to < 0 (returns from previous time windows)**

In [435]:
# Drop rows corresponding to returns and sales that would yield net qty < 0
all_data = all_data.drop(index = index_to_drop).reset_index(drop = True)

In [436]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 792443 entries, 0 to 792442
Data columns (total 9 columns):
InvoiceNo      792443 non-null int64
StockCode      792443 non-null object
Description    792443 non-null object
Quantity       792443 non-null int64
InvoiceDate    792443 non-null datetime64[ns]
UnitPrice      792443 non-null float64
CustomerID     792443 non-null int64
Country        792443 non-null object
Type           792443 non-null object
dtypes: datetime64[ns](1), float64(1), int64(3), object(4)
memory usage: 54.4+ MB


**Confirm all CustomerID/StockCode combos have quantities > 0**

In [439]:
assert all_data.groupby(['CustomerID','StockCode']).sum().min().all() >= 0

## 'InvoiceDate' Clean-Up

**Convert InvoiceDate to DateTime**

In [440]:
all_data['InvoiceDate'] = pd.to_datetime(all_data['InvoiceDate'])

In [441]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 792443 entries, 0 to 792442
Data columns (total 9 columns):
InvoiceNo      792443 non-null int64
StockCode      792443 non-null object
Description    792443 non-null object
Quantity       792443 non-null int64
InvoiceDate    792443 non-null datetime64[ns]
UnitPrice      792443 non-null float64
CustomerID     792443 non-null int64
Country        792443 non-null object
Type           792443 non-null object
dtypes: datetime64[ns](1), float64(1), int64(3), object(4)
memory usage: 54.4+ MB


# Save to CSV

In [442]:
all_data.to_csv('clean_data.csv')